File: content/problems/2503-maximum-number-of-points-from-grid-queries/solutions/offline-queries-with-frontier-heap/Solution.kt
class Solution {
    fun maxPoints(grid:Array<IntArray>,queries:IntArray):IntArray{
        val rows=grid.size;
        val cols=grid[0].size;
        val order=queries.indices.sortedBy{
            queries[it]
        };
        val heap=java.util.PriorityQueue<IntArray>(compareBy{
            it[0]
        });
        val visited=Array(rows){
            BooleanArray(cols)
        };
        heap.offer(intArrayOf(grid[0][0],0,0));
        visited[0][0]=true;
        val answer=IntArray(queries.size);
        var count=0;
        val directions=arrayOf(intArrayOf(1,0),intArrayOf(-1,0),intArrayOf(0,1),intArrayOf(0,-1));
        for(index in order){
            while(heap.isNotEmpty()&&heap.peek()[0]<queries[index]){
                val cell=heap.poll();
                count++;
                for(direction in directions){
                    val row=cell[1]+direction[0];
                    val col=cell[2]+direction[1];
                    if(row in 0 until rows&&col in 0 until cols&&!visited[row][col]){
                        visited[row][col]=true;
                        heap.offer(intArrayOf(grid[row][col],row,col))
                    }
                }
            };
            answer[index]=count
        };
        return answer
    }
}

File: content/problems/2503-maximum-number-of-points-from-grid-queries/solutions/offline-queries-with-frontier-heap/solution.cpp
class Solution {
    public:vector<int> maxPoints(vector<vector<int>>& grid,vector<int>& queries){
        int rows=grid.size(),cols=grid[0].size();
        vector<int> order(queries.size());
        iota(order.begin(),order.end(),0);
        sort(order.begin(),order.end(),[&](int a,int b){
            return queries[a]<queries[b];
        });
        using Cell=array<int,3>;
        priority_queue<Cell,vector<Cell>,greater<Cell>> heap;
        vector<vector<bool>> visited(rows,vector<bool>(cols));
        heap.push({
            grid[0][0],0,0
        });
        visited[0][0]=true;
        vector<int> answer(queries.size());
        int count=0,directions[5]={
            1,0,-1,0,1
        };
        for(int index:order){
            while(!heap.empty()&&heap.top()[0]<queries[index]){
                auto cell=heap.top();
                heap.pop();
                ++count;
                for(int direction=0;direction<4;++direction){
                    int row=cell[1]+directions[direction],col=cell[2]+directions[direction+1];
                    if(row>=0&&row<rows&&col>=0&&col<cols&&!visited[row][col]){
                        visited[row][col]=true;
                        heap.push({
                            grid[row][col],row,col
                        });
                    }
                }
            }
            answer[index]=count;
        }
        return answer;
    }
};
